### PR TITLE
Try to delete the virtualenv before creating it. fixes #1757

### LINKED ISSUE #1757: Reinstalling mmonit pack fails as the venv is being used
Im sure this is due the sensor being run, so you cant create the virtualenv again as its being used.

Should the packs action check if there is sensors and try to stop them before reinstalling a pack?

```
  "stderr": "st2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Setting up virtualenv for pack \"mmonit\"\nst2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Creating virtualenv for pack \"mmonit\" in \"/opt/stackstorm/virtualenvs/mmonit\"\nst2.actions.python.SetupVirtualEnvironmentAction: DEBUG    Creating virtualenv in \"/opt/stackstorm/virtualenvs/mmonit\" using Python binary \"/usr/bin/python2.7\"\nTraceback (most recent call last):\n  File \"/usr/lib/python2.7/dist-packages/st2actions/runners/python_action_wrapper.py\", line 116, in <module>\n    obj.run()\n  File \"/usr/lib/python2.7/dist-packages/st2actions/runners/python_action_wrapper.py\", line 61, in run\n    output = action.run(**self._parameters)\n  File \"/opt/stackstorm/packs/packs/actions/pack_mgmt/setup_virtualenv.py\", line 52, in run\n    self._setup_pack_virtualenv(pack_name=pack_name)\n  File \"/opt/stackstorm/packs/packs/actions/pack_mgmt/setup_virtualenv.py\", line 89, in _setup_pack_virtualenv\n    self._create_virtualenv(virtualenv_path=virtualenv_path)\n  File \"/opt/stackstorm/packs/packs/actions/pack_mgmt/setup_virtualenv.py\", line 125, in _create_virtualenv\n    (virtualenv_path, stderr))\nException: Failed to create virtualenv in \"/opt/stackstorm/virtualenvs/mmonit\": Traceback (most recent call last):\n  File \"/usr/local/lib/python2.7/dist-packages/virtualenv.py\", line 2363, in <module>\n    main()\n  File \"/usr/local/lib/python2.7/dist-packages/virtualenv.py\", line 832, in main\n    symlink=options.symlink)\n  File \"/usr/local/lib/python2.7/dist-packages/virtualenv.py\", line 994, in create_environment\n    site_packages=site_packages, clear=clear, symlink=symlink))\n  File \"/usr/local/lib/python2.7/dist-packages/virtualenv.py\", line 1288, in install_python\n    shutil.copyfile(executable, py_executable)\n  File \"/usr/lib/python2.7/shutil.py\", line 83, in copyfile\n    with open(dst, 'wb') as fdst:\nIOError: [Errno 26] Text file busy: '/opt/stackstorm/virtualenvs/mmonit/bin/python2.7'\n\n",
```

### PR DESCRIPTION
**Pull request by Itxaka**
Kind of a workaround as it would be better to stop the sensor, delete the venv, create it and register the sensor again so it cleaner.

On sensor registration, sensorcontainer kills the old one and launches the new one so there seems to be no side effects for this.

```
2015-07-29 21:43:35,644 140695903024464 INFO log [-] Container shutting down. Invoking cleanup on sensors.
2015-07-29 21:43:35,932 139779226957136 INFO log [-] Connecting to database "st2" @ "0.0.0.0:27017" as user "None".
2015-07-29 21:43:35,972 139779226957136 INFO log [-] Using partitioner default with sensornode sensornode1.
2015-07-29 21:43:35,973 139779226957136 INFO log [-] Found 2 registered sensors in db scan.
2015-07-29 21:43:35,973 139779226957136 INFO log [-] Setting up container to run 2 sensors.
2015-07-29 21:43:35,973 139779226957136 INFO log [-] 	Sensors list - [u'linux.FileWatchSensor', u'mmonit.MmonitEventsSensor'].
2015-07-29 21:43:35,974 139779226957136 INFO log [-] (PID:22506) SensorContainer started.
2015-07-29 21:43:35,974 57355632 INFO log [-] Running sensor linux.FileWatchSensor
2015-07-29 21:43:35,982 57355472 INFO mixins [-] Connected to amqp://guest:**@127.0.0.1:5672//
2015-07-29 21:43:35,985 57355632 AUDIT log [-] Access granted to "sensors_container" with the token set to expire at "2015-07-30T21:43:35.984435Z". (username='sensors_container',token_expiration='2015-07-30T21:43:35.984435Z')
2015-07-29 21:43:36,007 57355632 INFO log [-] Sensor linux.FileWatchSensor started
2015-07-29 21:43:36,007 57355632 INFO log [-] Running sensor mmonit.MmonitEventsSensor
2015-07-29 21:43:36,012 57355632 AUDIT log [-] Access granted to "sensors_container" with the token set to expire at "2015-07-30T21:43:36.011274Z". (username='sensors_container',token_expiration='2015-07-30T21:43:36.011274Z')
2015-07-29 21:43:36,028 57355632 INFO log [-] Sensor mmonit.MmonitEventsSensor started
2015-07-29 21:43:36,267 38551376 INFO log [-] Using config "/opt/stackstorm/packs/linux/config.yaml" for sensor "FileWatchSensor"
2015-07-29 21:43:36,304 22353552 INFO log [-] Using config "/opt/stackstorm/packs/mmonit/config.yaml" for sensor "MmonitEventsSensor"
2015-07-29 21:43:36,304 22353552 INFO log [-] Watcher started
2015-07-29 21:43:36,304 22353552 INFO log [-] Running sensor initialization code
2015-07-29 21:43:36,327 38299440 INFO mixins [-] Connected to amqp://guest:**@127.0.0.1:5672//
2015-07-29 21:43:36,362 22353552 INFO log [-] Running sensor in active mode (poll interval=10s)

```

### CODE REVIEW

**Comment by Itxaka** [comment]
Tested with mmonit pack: It wont fail like before.
Testes with other packs: Not affected

Tries to fix #1757

**Comment by manasdk** [comment]
@Itxaka  With this change new code will be present on disk but any new code will probably not be registered. Are you ok with this behavior in the short term?

**Comment by Itxaka** [comment]
@manasdk Umm, I do not understand that part. This is only for the virtualenv, so only the requirements will be removed, then they will be installed and the sensor will be registered and restarted (as per my tests)
Not sure in which part will any new code not being registered?

**Comment by manasdk** [comment]
@Itxaka  To install the new pack do you simply use `st2 run pack.install packs=my_pack` ?

Are there other steps you now follow. If there are other steps then likely you do not see this issue.

**Comment by Itxaka** [comment]
@manasdk Yup, only that:

```
st2 run packs.install packs=mmonit
id: 55b9ffd6e8b34d08511630af
action.ref: packs.install
status: succeeded
start_timestamp: 2015-07-30T10:43:34.341341Z
end_timestamp: 2015-07-30T10:43:46.821624Z
+--------------------------+-----------+--------------------------+-------------------------+-------------------------------+
| id                       | status    | task                     | action                  | start_timestamp               |
+--------------------------+-----------+--------------------------+-------------------------+-------------------------------+
| 55b9ffd6e8b34d084c2889b2 | succeeded | download                 | packs.download          | Thu, 30 Jul 2015 10:43:34 UTC |
| 55b9ffd9e8b34d084c2889b4 | succeeded | virtualenv_prerun        | packs.virtualenv_prerun | Thu, 30 Jul 2015 10:43:37 UTC |
| 55b9ffdae8b34d084c2889b6 | succeeded | setup_virtualenv         | packs.setup_virtualenv  | Thu, 30 Jul 2015 10:43:38 UTC |
| 55b9ffdee8b34d084c2889b8 | succeeded | reload                   | packs.load              | Thu, 30 Jul 2015 10:43:42 UTC |
| 55b9ffe0e8b34d084c2889ba | succeeded | restart-sensor-container | packs.restart_component | Thu, 30 Jul 2015 10:43:44 UTC |
+--------------------------+-----------+--------------------------+-------------------------+-------------------------------+

```

So I dont know where it could fail. Can you pinpoint it to me so I can try to fix any issues that could arise from this patch (and also so I can learn more about st2 internals :) )

thanks!

**Comment by manasdk** [comment]
@Itxaka Maybe I read the code wrong. I expected that the workflow would fail if `setup_virtuelenv` would throw an expection. I suspect this is really not any worse than what would happen right now. Let me pull your branch and try this out locally. I will report my results and we can take it from there :)

> Can you pinpoint it to me so I can try to fix any issues that could arise from this patch (and also so I can learn more about st2 internals :) )

Glad to see you wanting to learn more about st2 internals :)

**Comment by manasdk** [comment]
@Itxaka - I played with this locally and can now see that you are right. 

My expectation was that while a process is using the python interpreter from a specific virtualenv it woudl not allow the directory to be cleaned-up. Of course that looks like fair game and likely coz the virtualenv anyway only contains symlink to system python binaries. My guess is that if the interpreter tries to load some modules from the virtualenv while the env is not created it is likely to blow-up. But this is probably a small enough window anyway.

Long story short ... works as you described and I learned something new and surprising.

+1

**Comment by Itxaka** [comment]
@manasdk indeed I expected the same behavior and was surprised when it worked, and basically arrived at the same conclusion as yours. 

Thanks!
